### PR TITLE
[FEATURE SERVERSIDE APPLY] Fallback to client-side apply if incompatible server

### PR DESCRIPTION
**Pull request by seans3**
Fallback from server-side apply to client-side apply if we detect we're talking to a server which does not support server-side apply.

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/cmd/apply.go [hunk 1]
```diff
@@ -340,35 +347,32 @@ func (o *ApplyOptions) Run() error {
...
 			}
-			// TODO: If we get a 415 (unsupported media type), then the server doesn't understand
-			// serverside apply; continue flow after this block.
 			obj, err := resource.NewHelper(info.Client, info.Mapping).
 				DryRun(o.DryRun).
 				Patch(info.Namespace, info.Name, types.ApplyPatchType, data)
-			if err != nil {
+			// If we're talking to a server which does not implement server-side apply,
+			// continue with the client side apply after this block.
+			if err != nil && !isIncompatibleServerError(err) {
```
Comment (apelisse) [inline pkg/kubectl/cmd/apply.go]: Code is fine. 

Have you consider this alternative structure:
```
if err == nil {
    // carry-on with server-side apply
    return nil
} else if !isIncompatibleServerError(err) {
    return err
}
fallback_to_regular_apply
```

WDYT?

File: pkg/kubectl/cmd/apply.go [hunk 2]
```diff
@@ -282,6 +283,12 @@ func parsePruneResources(mapper meta.RESTMapper, gvks []string) ([]pruneResource
 	return pruneResources, nil
 }
 
+func isIncompatibleServerError(err error) bool {
+	// 415: Unsupported media type means we're talking to a server which doesn't
+	// support server-side apply.
+	return err.(*errors.StatusError).Status().Code == http.StatusUnsupportedMediaType
```
Comment (apelisse) [inline pkg/kubectl/cmd/apply.go]: Won't this crash if the error isn't of type `*errors.StatusError`?

Comment (seans3) [reply pkg/kubectl/cmd/apply.go]: The error returned is always a StatusError.

Comment (seans3) [reply pkg/kubectl/cmd/apply.go]: After our discussion, I agree we should be more defensive and check if it's a StatusError first.

**Comment by apelisse** [comment]
LGTM, please squash

**Comment by apelisse** [comment]
That's great news!

**Pull request merged**